apps/web/pages/api/book/event.ts

```diff
@@ -17,8 +25,31 @@ async function handler(req: NextApiRequest & { userId?: number }, res: NextApiRe
   const session = await getServerSession({ req, res });
   /* To mimic API behavior and comply with types */
   req.userId = session?.user?.id || -1;
-  const booking = await handleNewBooking(req);
-  return booking;
+
+  const requestBody = JSON.stringify({ ...req.body, userIp });
+  const idempotencyKey = generateHash(requestBody);
+
+  try {
+    const checkIfKeyExist = await redis.exists(idempotencyKey);
+    if (checkIfKeyExist) {
+      return res.status(429).json({ message: "Request Already Processing" });
+    }
+
+    await redis.set(idempotencyKey, "1");
+    const booking = await handleNewBooking(req);
+    await redis.del(idempotencyKey);
+
+    return booking;
+  } catch (err: unknown) {
+    // Extract error information
+    const responseData = { message: err?.message ?? "Internal Server Error" };
+    const statusCode = err?.statusCode ?? 500;
+
+    await redis.del(idempotencyKey);
```

Comment:
Nit: can just push the redis.del in a finally block so we don’t have this line twice